primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -910,4 +1247,122 @@ mod tests {
+        fn gpu_profiling<E, C>() -> Result<(), PCSError>
+        where
+            C: IcicleCurve + MSM<C>,
+            C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+            C::BaseField:
+                ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+            <C::ArkSWConfig as CurveConfig>::BaseField: PrimeField,
+            E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+            UnivariateKzgPCS<E>: GPUCommit<E, C>,
+        {
+            let rng = &mut test_rng();
+            let stream = warmup_new_stream().unwrap();
+            let degree = 2usize.pow(20);
+
+            let pp = UnivariateKzgPCS::<E>::gen_srs_for_testing(rng, degree)?;
+            let (ck, _vk) = pp.trim(degree)?;
+            let mut srs_on_gpu =
+                <UnivariateKzgPCS<E> as GPUCommit<E, C>>::load_prover_param_to_gpu(&ck, degree)?;
+
+            let p = <DensePolynomial<E::ScalarField> as DenseUVPolynomial<E::ScalarField>>::rand(
+                degree, rng,
+            );
+            let comm = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::commit_with_gpu(&ck, &p)?;
+
+            // step-by-step commitment on gpu
+            let poly_on_gpu = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::load_poly_to_gpu(&p)?;
+            let msm_result_on_gpu = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::commit_on_gpu(
+                &mut srs_on_gpu,
+                &poly_on_gpu,
+                1,
+                &stream,
+            )?;
+            let comm2 = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::load_commitments_to_host(
+                msm_result_on_gpu,
+                &stream,
+            )?[0];
+            assert_eq!(comm, comm2);
+
+            let polys: Vec<_> = (0..10)
+                .map(|_| {
+                    <DensePolynomial<E::ScalarField> as DenseUVPolynomial<E::ScalarField>>::rand(
+                        degree, rng,
+                    )
+                })
+                .collect();
+            let _comms =
+                <UnivariateKzgPCS<E> as GPUCommit<E, C>>::batch_commit_with_gpu(&ck, &polys)?;
+
+            Ok(())
+        }
+
+        fn test_gpu_e2e_template<E, C>() -> Result<(), PCSError>
+        where
+            C: IcicleCurve + MSM<C>,
+            C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+            C::BaseField:
+                ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+            <C::ArkSWConfig as CurveConfig>::BaseField: PrimeField,
+            E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+            UnivariateKzgPCS<E>: GPUCommit<E, C>,
+        {
+            let rng = &mut test_rng();
+            let supported_degree = 2usize.pow(12);
+            let pp = UnivariateKzgPCS::<E>::gen_srs_for_testing(rng, supported_degree)?;
+
+            // testing on smaller degree for correctness
+            for _ in 0..10 {
```

Comment:
weird thing. My local test is failing if I change it to 0..50

Reply:
ofc, 2^20 * 50 * 32, your GPU memory exhausted

Reply:
I thought they'll be freed. NVM

Reply:
🤔 you have a good point, lemme investigate why it's not dropping.

Reply:
here's what I observe so far by watching GPU memory usage:

- at startup, before I load anything onto it, as long as i create a cuda stream, `260 MB` of memory will be used
- after SRS is loaded, another 66 MB (which is expected for 2^20) will be used
- after a single commit, at the end of it, there's another 2MB dangling extra usage (I expect everything should be cleaned except the SRS). apparently scalars are freed, cuz they will be 32 MB if not, the MSM result is just ~100B, and I have no clue so far why there's extra 2MB increase. Either somewhere in my code or ICICLE MSM code, we fail to clean up.

Reply:
This is just # of iterations not batch_size right? And the degree below is nowhere near 2^20.